test_nnf.py

```diff
@@ -754,3 +754,24 @@ def test_dsharp_compile_converting_names(sentence: And[Or[Var]]):
         assert all(isinstance(name, str) for name in compiled.vars())
         if sentence.satisfiable():
             assert sentence.equivalent(compiled)
+
+
+@given(NNF())
+def test_tseitin(sentence: nnf.NNF):
+    sentence = sentence.simplify()
+    assume(len(sentence.vars()) <= 6)
+    assume(len(sentence.vars()) >= 2)
+
+    T = sentence.simplify()
+    T = tseitin.to_cnf(T)
+
+    # TODO: Once forgetting/projection is implemented, do this more complete check
+    # aux = filter(lambda x: 'aux' in str(x.name), T.vars())
+    # assert T.forget(aux).equivalent(sentence)
+
+    if not (T in [true, false] or isinstance(T, Var)):
+        for mt in T.models():
+            ms = {k:v for (k,v) in mt.items() if k in sentence.vars()}
+            assert sentence.satisfied_by(ms)
+
+    assert T.model_count() == sentence.model_count()
```

Comment:
I'd also `assert T.is_CNF()`.

Reply:
Issue here (without simplification) is that any CNF with a clause that has a variable+it's negation is deemed non-simple (and thus not a clause). So adding this assert breaks on the test cause `x & ~x`

Reply:
Ah, right. That's a shame. A few possibilities:

- Change the behavior of `is_CNF()` (and possibly `is_DNF()`). I don't know how standard the definition in the KCM is. Would users of the library expect them to be so strict?
Dropping the requirement for `is_DNF()` would mean that DNF is no longer a subset of DNNF.
Dropping the requirement for `is_CNF()` but not `is_DNF()` would mean that not all negated CNF sentences are DNF.
- Change `to_CNF()` to not emit those clauses because they're trivially true. And then you might as well short-circuit and only emit a single clause `Or({~aux})` or `Or({aux})` for the node depending on whether it's an `And` or an `Or`.
Could be undesirable if the goal is to have an extremely straightforward and faithful Tseitin implementation, but not if the goal is to plug into SAT solvers.
- Both of the above.
- None of the above.

What do you think?

Reply:
Fun. Well if we take the KCM as mandated, then `is_DNF()` should remain, and as should `is_CNF()` because of your reasoning. In terms of these variants on theory restrictions, there's no source more adopted than the original KCM anyways, so it's a safe bet to adhere by.

I think it would be far rarer someone wants a truly faithful Tseitin rewrite, and maybe an optional parameter (like `skip_simplification` or whatever) could be used to keep things pure.

Reply:
With the minor simplification you recommended, we can now safely check `is_CNF()`. It comes at the price of not checking the model equivalence when some variables are pruned, but that will come back into play with (#11 ).